File: freamon-collector/src/java/de/tuberlin/cit/freamon/cgroups/YarnConfig.java
package de.tuberlin.cit.freamon.cgroups;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

/**
 * Helper for retrieving certain settings from yarn-site.xml
 */
public class YarnConfig {

	public final Map<String, String> configMap = new HashMap<String, String>();
	public String cgroupsMountPath;
	public String cgroupsHierarchy;
	public String configPath;

	public YarnConfig(String configPath) throws IOException {
		// TODO infer configPath from `HADOOP_CONF_DIR` or `HADOOP_PREFIX`
		this.configPath = configPath;

		try {
			ConfigurationUtils.load(configMap, this.configPath);
		} catch (IOException e) {
			throw new IOException("Could not load " + this.configPath);
		}

		this.cgroupsMountPath = getOrDefault("yarn.nodemanager.linux-container-executor.cgroups.mount-path", "/sys/fs/cgroup");
		this.cgroupsHierarchy = getOrDefault("yarn.nodemanager.linux-container-executor.cgroups.hierarchy", "hadoop-yarn");
	}

	public String getOrDefault(String key, String defaultValue) {
		// Java 8 Map has this method built in
		String value = configMap.get(key);
		if (value == null) {
			return defaultValue;
		}
		return value;
	}

}

File: freamon-collector/src/scala/de/tuberlin/cit/freamon/collector/YarnConfig.scala
package de.tuberlin.cit.freamon.collector

import java.io.IOException

/**
 * Helper for retrieving certain settings from yarn-site.xml
 */
class YarnConfig {
  var configPath: String = null
  var cgroupsMountPath: String = null
  var cgroupsHierarchy: String = null

  def this(configPath: String) {
    this()

    try {
      val file = scala.xml.XML.loadFile(configPath)
      val properties = file.child.filter(_.label.toLowerCase == "property").map { property =>
        val key = property.child.find(_.label.toLowerCase == "name").get.text
        val value = property.child.find(_.label.toLowerCase == "value").get.text
        (key, value)
      }.toMap

      this.cgroupsMountPath = properties.getOrElse("yarn.nodemanager.linux-container-executor.cgroups.mount-path", "/sys/fs/cgroup")
      this.cgroupsHierarchy = properties.getOrElse("yarn.nodemanager.linux-container-executor.cgroups.hierarchy", "hadoop-yarn")

      this.configPath = configPath
    }
    catch {
      case e: IOException =>
        throw new IOException("Could not load file: " + configPath, e)
      case e: Throwable =>
        throw new IOException("Wrong format: " + configPath, e)
    }
  }

}
